frappe/migrate.py

```diff
@@ -35,70 +59,117 @@ def migrate(verbose=True, skip_failing=False, skip_search_index=False):
 		frappe.model.sync.sync_all()
-		frappe.modules.patch_handler.run_all(skip_failing=skip_failing, patch_type=PatchType.post_model_sync)
-		frappe.translate.clear_cache()
+		frappe.modules.patch_handler.run_all(skip_failing=self.skip_failing, patch_type=PatchType.post_model_sync)
+
+	@atomic
+	def post_schema_updates(self):
+		"""Execute pending migration tasks post patches execution & schema sync
+		This includes:
+		* Sync `Scheduled Job Type` and scheduler events defined in hooks
+		* Sync fixtures & custom scripts
+		* Sync in-Desk Module Dashboards
+		* Sync customizations: Custom Fields, Property Setters, Custom Permissions
+		* Sync Frappe's internal language master
+		* Sync Portal Menu Items
+		* Sync Installed Applications Version History
+		* Execute `after_migrate` hooks
+		"""
 		sync_jobs()
 		sync_fixtures()
 		sync_dashboards()
 		sync_customizations()
 		sync_languages()
 
-		frappe.get_doc('Portal Settings', 'Portal Settings').sync_menu()
-
-		# syncs static files
-		clear_website_cache()
-
-		# updating installed applications data
-		frappe.get_single('Installed Applications').update_versions()
+		frappe.get_single("Portal Settings").sync_menu()
+		frappe.get_single("Installed Applications").update_versions()
 
 		for app in frappe.get_installed_apps():
-			for fn in frappe.get_hooks('after_migrate', app_name=app):
+			for fn in frappe.get_hooks("after_migrate", app_name=app):
 				frappe.get_attr(fn)()
 
-		if not skip_search_index:
-			# Run this last as it updates the current session
-			print('Building search index for {}'.format(frappe.local.site))
-			build_index_for_all_routes()
-
-		frappe.db.commit()
-
-		clear_notifications()
-
-		frappe.publish_realtime("version-update")
-		frappe.flags.in_migrate = False
-	finally:
-		with open(touched_tables_file, 'w') as f:
-			json.dump(list(frappe.flags.touched_tables), f, sort_keys=True, indent=4)
-		frappe.flags.touched_tables.clear()
+	def required_services_running(self) -> bool:
+		"""Returns True if all required services are running. Returns False and prints
+		instructions to stdout when required services are not available.
+		"""
+		service_status = check_connection(redis_services=["redis_cache"])
+		are_services_running = all(service_status.values())
+
+		if not are_services_running:
+			for service in service_status:
+				if not service_status.get(service, True):
+					print(f"Service {service} is not running.")
+			print(BENCH_START_MESSAGE)
+
+		return are_services_running
+
+	def run(self, site: str):
+		"""Run Migrate operation on site specified. This method initializes
+		and destroys connections to the site database.
+		"""
+		if not self.required_services_running():
+			raise SystemExit(1)
+
+		if site:
+			frappe.init(site=site)
+			frappe.connect()
+
+		self.setUp()
+		try:
+			self.pre_schema_updates()
+			self.run_schema_updates()
+		finally:
+			self.post_schema_updates()
+			self.tearDown()
+			frappe.destroy()
```

Comment:
```suggestion
		self.setUp()
		try:
			self.pre_schema_updates()
			self.run_schema_updates()
			self.post_schema_updates()
		finally:
			self.tearDown()
			frappe.destroy()
```

Wondering if the finally block should only handle tearDown & destroying locals & connections since `self.post_schema_updates` executes 

* Sync `Scheduled Job Type` and scheduler events defined in hooks
* Sync fixtures & custom scripts
* Sync in-Desk Module Dashboards
* Sync customizations: Custom Fields, Property Setters, Custom Permissions
* Sync Frappe's internal language master
* Sync Portal Menu Items
* Sync Installed Applications Version History
* **Execute `after_migrate` hooks**

---

Although I think it's cool -  don't see anything catastrophic happening

cc: @frappe/frappe-review-team @frappe/community-contributors